helix-core/src/date.rs

```diff
@@ -0,0 +1,217 @@
+use chrono::{Duration, NaiveDate};
+
+use std::borrow::Cow;
+
+use ropey::RopeSlice;
+
+use crate::{
+    textobject::{textobject_word, TextObject},
+    Range, Tendril,
+};
+
+// Only support formats that aren't region specific.
+static FORMATS: &[&str] = &["%Y-%m-%d", "%Y/%m/%d"];
+
+// We don't want to parse ambiguous dates like 10/11/12 or 7/8/10.
+// They must be YYYY-mm-dd or YYYY/mm/dd.
+// So 2021-01-05 works, but 2021-1-5 doesn't.
+const DATE_LENGTH: usize = 10;
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct DateIncrementor {
+    pub date: NaiveDate,
+    pub range: Range,
+    pub format: &'static str,
+}
+
+impl DateIncrementor {
+    pub fn from_range(text: RopeSlice, range: Range) -> Option<DateIncrementor> {
+        // Don't increment if the cursor is one right of the date text.
+        if text.char(range.from()).is_whitespace() {
+            return None;
+        }
+
+        let range = textobject_word(text, range, TextObject::Inside, 1, true);
+        let text: Cow<str> = text.slice(range.from()..range.to()).into();
+
+        let first = text.chars().next()?;
+        let last = text.chars().next_back()?;
+
+        // Allow date strings in quotes.
+        let (range, text) = if first == last && (first == '"' || first == '\'') {
+            (
+                Range::new(range.from() + 1, range.to() - 1),
+                Cow::from(&text[1..text.len() - 1]),
+            )
+        } else {
+            (range, text)
+        };
```

Comment:
I think this solution might be very limited say if the big word selected `'date',` or `"date";` or other cases, my suggestion solution that only works with cursor head or you can fallback to yours when it select the whole date.

1. check if the cursor head is a digit
2. use regular expression to search date (within 10 radius from the current position)
3. check if the cursor is within any of the matches
4.

Reply:
Yeah, this is definitely too limiting as it is now. The approach you're suggesting should work well I think.

However, what I do here will depend on whether we have a good way to increment the year and month fields. As it is now, the increment will work if the cursor head is on a `-` or a `/`, but it only increments days.